app/views.py

```diff
@@ -0,0 +1,242 @@
+                email=user_data['email'], 
+                password=user_data['password'], 
+                username=user_data['username'], 
+                phone_no=user_data['phone_no'],
+                date_of_birth=user_data['date_of_birth'])
+
+            user = authenticate(email=user_data['email'], password=user_data['password'])
+            token, _ = Token.objects.get_or_create(user=user)
+            user_data['token'] = token.key
+            return Response({"message":"User Signed up successfully", "User":user_data}, status=status.HTTP_201_CREATED)
+        else:
+            return Response({"message":"Invalid Data"}, status=status.HTTP_400_BAD_REQUEST)
+
+    # Check if user exists or not
+    def get(self, request):
+        email = request.query_params.get("email")
+        password = request.query_params.get("password")
+        print(email + "___" + password)
+        user = authenticate(email=email, password=password)
+        if not user:
+            return Response({"message":"User does not exist"}, status=status.HTTP_204_NO_CONTENT)
+        else:
+            return Response({"message":"User Already Exists"}, status=status.HTTP_302_FOUND)
+
+# View for user login
+class UserLoginView(APIView):
+    permission_classes = (AllowAny,)
+    
+    def post(self, request):
+        req_data = request.data
+        user = authenticate(email=req_data['email'], password=req_data['password'])
+        if not user:
+            return Response({"message":"Invalid Details"}, status=status.HTTP_400_BAD_REQUEST)
+        else:
+            token, _ = Token.objects.get_or_create(user=user)
+            return Response({
+                "message":"User Logged In", 
+                "User":{
+                    "id":user.id,
+                    "email":user.email,
+                    "username":user.username,
+                    "phone_no":user.phone_no,
+                    "date_of_birth":user.date_of_birth,
+                    "token":token.key
+            }})
+
+# Signout new user
+class UserLogoutView(APIView):
+    permission_classes = (IsAuthenticated,)
+
+    def get(self, request, format=None):
+        user = request.user
+        response = {
+            "message":"User logged out", 
+            "Details":{
+                "id": user.id,
+                "email":user.email,
+                "username":user.username,
+                "phone_no":user.phone_no,
+                "date_of_birth":user.date_of_birth
+            }}
+        request.user.auth_token.delete()
+        return Response(response, status=status.HTTP_200_OK)
+
+# Register a new device to backend and store registration_id
+class FCMRegisterDeviceView(APIView):
+    permission_classes = (IsAuthenticated,)
+
+    # Register a new device (create new object)
+    def post(self, request):
+        req_data = request.data
+        user = request.user
+        try:
+            device = FCMDevice.objects.get(user=user)
```

Comment:
FCMDevice is a local model right? Here inside of try and except block condition should be checked whether the user exists or none or not. Because no exception is thrown if the user is found or not. So the except block will not be executed.

Reply:
FCMDevice is a model on the database itself. I have used directly fcm-django which makes this model directly on performing migrations. There might be a user who does not have a registered have a registered device. (A device stored corresponding to that user). In that case except part will be executed